Export typescript types from index.d.ts

fschucht opened the pull request:
Right now individual types are not being exported the `index.d.ts` file so they can't be reused outside of this module. For example in our use case we would need to access the `WSChain` for a utility function which wraps `superwstest` which fails with the current types:

```ts
import wsRequest, { WSChain } from 'superwstest'; // WSChain is not exported and can't be reused

export async function createApiWebsocketConnection(
  app: INestApplication,
  route: string,
  authToken?: string,
): Promise<{ connection: WSChain }> {
  const headers = {};

  if (authToken) {
    headers['Authorization'] = `Bearer ${authToken}`;
  }

  const randomPort = await getPort();
  await app.listen(randomPort);

  const connection = wsRequest(app.getHttpServer())
    .ws(`${route}/subscriptions`, ['graphql-ws'])
    .sendJson({
      payload: headers,
      type: GraphQlSubscriptionType.CONNECTION_INIT,
    })
    .expectJson({
      type: GraphQlSubscriptionType.CONNECTION_ACKNOWLEDGED,
    });

  return { connection };
}
```

This pr attempts to fix this by exporting the types in addition to the default export. It works for us in combination with typescript `4.4.2`. It also adds a `.eslintignore` file which ignores the `index.d.ts` file as it was wrongly complaining about type declarations  with the following error: `Parsing error: Imports within a 'declare module' body must always be 'import type' or 'import typeof'.`

davidje13 commented:
I'm not sure why the linter's running on that file for you in the first place (it's configured to only run for `.js` / `.mjs` / `.jsx` files), but I think it's actually correct; all the imports in that file ought to be `import type`:

```javascript
  import type { Server, ClientRequestArgs, IncomingMessage } from 'http';
  import type WebSocket from 'ws';
  import type { SuperTest, Test } from 'supertest';
```

Could you fix that instead of adding an ignore file? (I don't think `import type` was a thing back when I started this, but now it's pretty well-established and I see no reason to maintain backwards compatibility with such old TypeScript versions)

As for exporting the types: sounds good, except I'd remove the exports for the JSON-type values (`JsonObject` / `JsonArray` / `JsonValue`) — I don't think anyone would need those types and they're not really connected to what this library does; just needed because TypeScript doesn't have a built-in JSON type (let me know if you have a particular need for them but I can't think of anything). The others all make sense.

fschucht commented:
Thanks for the input. I updated the pr to remove the `.eslintignore` file again and do the `import type` instead. I also removed the json type exports.

> except I'd remove the exports for the JSON-type values (JsonObject / JsonArray / JsonValue) — I don't think anyone would need those types and they're not really connected to what this library does; just needed because TypeScript doesn't have a built-in JSON type 

We don't have a need for those types. The underlying thought was to export them in order to keep the library open in case someone needs them. But they are also quite basic types so there is not necessarily a need for that I think.

davidje13 commented:
Merged and published as 1.5.1. Thanks for the contribution :)
